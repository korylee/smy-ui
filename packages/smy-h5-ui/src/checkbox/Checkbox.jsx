import { assign } from '../_utils/shared'
import { createNamespace } from '../_utils/vue/create'
import Checker from './Checker'
import { props } from './props'
import { createChildrenMixin } from '../_mixins/relation'

import '../_styles/common.less'
import './checkbox.less'

const [name, bem] = createNamespace('checkbox')

export default {
  name,
  mixins: [createChildrenMixin('checkboxGroup')],
  props,
  model: {
    prop: 'checked',
    event: 'input',
  },
  computed: {
    _checked({ checked, checkboxGroup, bindGroup, value }) {
      if (checkboxGroup && bindGroup) {
        return checkboxGroup.value.indexOf(value) !== -1
      }
      return !!checked
    },
  },
  watch: {
    value(newValue) {
      if (this.indeterminate === null) {
        this.$emit('change', newValue)
      }
    },
  },
  methods: {
    toggle(newValue = !this._checked) {
      const { checkboxGroup: parent, bindGroup, value } = this
      if (parent && bindGroup) {
        const { max, value: parentValue } = parent.$props
        const values = parentValue.slice()
        if (newValue) {
          const overlimit = max && values.length >= +max
          if (!overlimit && !values.includes(value)) {
            values.push(value)
            parent.updateValue(values)
          }
        } else {
          const index = values.indexOf(value)
          if (index !== -1) {
            values.splice(index, 1)

            parent.updateValue(values)
          }
        }
      } else {
        this.$emit('input', newValue)
      }
      if (this.indeterminate !== null) {
        this.$emit('change', newValue)
      }
    },
  },
  render(h) {
    const vm = this
    const scopedSlots = assign({}, vm.$scopedSlots)

    const { _checked: checked, checkboxGroup: parent } = vm
    const attrs = assign({}, vm.$props, {
      role: 'checkbox',
      bem,
      checked,
      parent,
    })
    return h(Checker, {
      attrs,
      on: { toggle: vm.toggle },
      scopedSlots,
    })
  },
}
