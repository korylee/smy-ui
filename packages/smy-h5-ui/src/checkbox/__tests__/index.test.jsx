import { mount } from '@vue/test-utils'
import Checkbox from '..'

test('should emit "input" event when checkbox icon is checked', async () => {
  const wrapper = mount(Checkbox)

  const icon = wrapper.find('.smy-checkbox__icon')
  icon.trigger('click')
  expect(wrapper.emitted('input')[0]).toEqual([true])

  await wrapper.setProps({ checked: true })
  icon.trigger('click')
  expect(wrapper.emitted('input')[1]).toEqual([false])
})

test('should emit change event when checked is changed', async () => {
  const wrapper = mount(Checkbox)

  const icon = wrapper.find('.smy-checkbox__icon')
  icon.trigger('click')
  await wrapper.setProps({ checked: true })
  expect(wrapper.emitted('change')[0]).toEqual([true])

  icon.trigger('click')
  await wrapper.setProps({ checked: false })
  expect(wrapper.emitted('change')[1]).toEqual([false])
})

test('should render "smy-checkbox--label-disabled" class when using label-disabled prop', async () => {
  const wrapper = mount(Checkbox, {
    propsData: {
      labelDisabled: true,
    },
    slots: {
      default: 'label',
    },
  })
  expect(wrapper.classes()).toContain('smy-checkbox--label-disabled')
  const label = wrapper.find('.smy-checkbox__label')
  label.trigger('click')
  expect(wrapper.emitted('input')).toBeFalsy()

  await wrapper.setProps({
    labelDisabled: false,
  })
  await label.trigger('click')
  expect(wrapper.emitted('input')).toBeTruthy()
})

test('should adjust label position when using label-position prop', () => {
  const wrapper = mount(Checkbox, {
    propsData: {
      labelPosition: 'left',
    },
    slots: {
      default: 'label',
    },
  })
  expect(wrapper.html()).toMatchSnapshot()
})
test('should emit click event when checkbox icon is clicked', async () => {
  const onClick = jest.fn()
  const wrapper = mount(Checkbox, {
    listeners: {
      click: onClick,
    },
  })
  wrapper.trigger('click')
  expect(onClick).toHaveBeenCalledTimes(1)

  const icon = wrapper.find('.smy-checkbox__icon')
  icon.trigger('click')
  expect(onClick).toHaveBeenCalledTimes(2)
})

test('should render icon & label slot correctly', async () => {
  const wrapper = mount(Checkbox, {
    scopedSlots: {
      icon: ({ checked, disabled }) => `checked: ${checked}, disabled: ${disabled}`,
    },
    slots: {
      default: 'Custom label',
    },
  })
  expect(wrapper.html()).toMatchSnapshot()
})

test('should render label slot correctly', async () => {
  const slot = jest.fn()
  const wrapper = mount(Checkbox, {
    scopedSlots: {
      default: slot,
    },
  })
  expect(slot.mock.calls[0]).toEqual([{ checked: false, disabled: false }])

  await wrapper.setProps({ checked: true })
  expect(slot.mock.calls[1]).toEqual([{ checked: true, disabled: false }])

  await wrapper.setProps({ disabled: true })
  expect(slot.mock.calls[2]).toEqual([{ checked: true, disabled: true }])
})
