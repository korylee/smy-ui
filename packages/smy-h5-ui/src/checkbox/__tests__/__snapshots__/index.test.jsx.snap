// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should adjust label position when using label-position prop 1`] = `
"<div role="checkbox" tabindex="0" class="smy-checkbox"><span class="smy-checkbox__label smy-checkbox__label--left">label</span>
  <div class="smy-checkbox__icon smy-checkbox__icon--square smy-checkbox__icon--preset"><span class="smy-icon smy-checkbox__icon-content" style="transition-duration: 0ms;"><svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="24" height="24" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></div>
</div>"
`;

exports[`should render icon & label slot correctly 1`] = `
"<div role="checkbox" tabindex="0" class="smy-checkbox">
  <div class="smy-checkbox__icon smy-checkbox__icon--square smy-checkbox__icon--preset">checked: false, disabled: false</div><span class="smy-checkbox__label">Custom label</span>
</div>"
`;
